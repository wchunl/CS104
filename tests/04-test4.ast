ROOT "" (0.0.0)
|   TOK_FUNCTION "" (6.22.1) function
|   |   TOK_TYPE_ID "" (6.22.1) typeid
|   |   |   TOK_VOID "void" (6.22.1) void
|   |   |   TOK_IDENT "fail" (6.22.6) variable
|   |   TOK_PARAM "(" (6.22.11) param
|   |   |   TOK_TYPE_ID "" (6.22.12) typeid
|   |   |   |   TOK_STRING "string" (6.22.12) string
|   |   |   |   TOK_IDENT "expr" (6.22.19) string variable (6.22.12)
|   |   |   TOK_TYPE_ID "" (6.22.25) typeid
|   |   |   |   TOK_STRING "string" (6.22.25) string
|   |   |   |   TOK_IDENT "file" (6.22.32) string variable (6.22.25)
|   |   |   TOK_TYPE_ID "" (6.22.38) typeid
|   |   |   |   TOK_INT "int" (6.22.38) int
|   |   |   |   TOK_IDENT "line" (6.22.42) int variable (6.22.38)
|   TOK_FUNCTION "" (6.24.1) function
|   |   TOK_TYPE_ID "" (6.24.1) typeid
|   |   |   TOK_VOID "void" (6.24.1) void
|   |   |   TOK_IDENT "putchr" (6.24.6) variable
|   |   TOK_PARAM "(" (6.24.13) param
|   |   |   TOK_TYPE_ID "" (6.24.14) typeid
|   |   |   |   TOK_INT "int" (6.24.14) int
|   |   |   |   TOK_IDENT "chr" (6.24.18) int variable (6.24.14)
|   TOK_FUNCTION "" (6.25.1) function
|   |   TOK_TYPE_ID "" (6.25.1) typeid
|   |   |   TOK_VOID "void" (6.25.1) void
|   |   |   TOK_IDENT "putint" (6.25.6) variable
|   |   TOK_PARAM "(" (6.25.13) param
|   |   |   TOK_TYPE_ID "" (6.25.14) typeid
|   |   |   |   TOK_INT "int" (6.25.14) int
|   |   |   |   TOK_IDENT "num" (6.25.18) int variable (6.25.14)
|   TOK_FUNCTION "" (6.26.1) function
|   |   TOK_TYPE_ID "" (6.26.1) typeid
|   |   |   TOK_VOID "void" (6.26.1) void
|   |   |   TOK_IDENT "putstr" (6.26.6) variable
|   |   TOK_PARAM "(" (6.26.13) param
|   |   |   TOK_TYPE_ID "" (6.26.14) typeid
|   |   |   |   TOK_STRING "string" (6.26.14) string
|   |   |   |   TOK_IDENT "str" (6.26.21) string variable (6.26.14)
|   TOK_FUNCTION "" (6.28.1) function
|   |   TOK_TYPE_ID "" (6.28.1) typeid
|   |   |   TOK_INT "int" (6.28.1) int
|   |   |   TOK_IDENT "getchr" (6.28.5) variable
|   |   TOK_PARAM "(" (6.28.11) param
|   TOK_FUNCTION "" (6.29.1) function
|   |   TOK_TYPE_ID "" (6.29.1) typeid
|   |   |   TOK_STRING "string" (6.29.1) string
|   |   |   TOK_IDENT "getstr" (6.29.8) variable
|   |   TOK_PARAM "(" (6.29.14) param
|   TOK_FUNCTION "" (6.30.1) function
|   |   TOK_TYPE_ID "" (6.30.1) typeid
|   |   |   TOK_STRING "string" (6.30.1) string
|   |   |   TOK_IDENT "getln" (6.30.8) variable
|   |   TOK_PARAM "(" (6.30.13) param
|   TOK_STRUCT "struct" (7.5.1)
|   |   TOK_IDENT "foo" (7.5.8) variable
|   |   TOK_INT "int" (7.6.4) int
|   |   |   TOK_TYPE_ID "a" (7.6.8) typeid
|   TOK_FUNCTION "" (7.9.1) function
|   |   TOK_TYPE_ID "" (7.9.1) typeid
|   |   |   TOK_INT "int" (7.9.1) int
|   |   |   TOK_IDENT "main" (7.9.5) variable
|   |   TOK_PARAM "(" (7.9.9) param
|   |   TOK_BLOCK "{" (7.9.12)
|   |   |   TOK_VARDECL "=" (7.10.10) variable lval
|   |   |   |   TOK_INT "int" (7.10.4) int
|   |   |   |   TOK_IDENT "a" (7.10.8) int variable (7.10.10)
|   |   |   |   TOK_INTCON "6" (7.10.12) int const
|   |   |   TOK_VARDECL "=" (7.11.22) variable lval
|   |   |   |   TOK_PTR "ptr" (7.11.4)
|   |   |   |   |   TOK_TYPE_ID "foo" (7.11.15) typeid
|   |   |   |   TOK_IDENT "b" (7.11.20) struct variable (7.11.22)
|   |   |   |   TOK_ALLOC "alloc" (7.11.24)
|   |   |   |   |   TOK_TYPE_ID "foo" (7.11.37) typeid
|   |   |   '=' "=" (7.12.9) int vreg
|   |   |   |   TOK_ARROW "->" (7.12.5) lval vaddr
|   |   |   |   |   TOK_IDENT "b" (7.12.4) struct variable (7.11.22)
|   |   |   |   |   TOK_IDENT "a" (7.12.7) int variable (7.10.10)
|   |   |   |   TOK_INTCON "8" (7.12.11) int const
|   |   |   '=' "=" (7.13.6) int vreg
|   |   |   |   TOK_IDENT "a" (7.13.4) int variable (7.10.10)
|   |   |   |   '+' "+" (7.13.17) int vreg
|   |   |   |   |   '*' "*" (7.13.10) int vreg
|   |   |   |   |   |   TOK_IDENT "a" (7.13.8) int variable (7.10.10)
|   |   |   |   |   |   TOK_ARROW "->" (7.13.13) lval vaddr
|   |   |   |   |   |   |   TOK_IDENT "b" (7.13.12) struct variable (7.11.22)
|   |   |   |   |   |   |   TOK_IDENT "a" (7.13.15) int variable (7.10.10)
|   |   |   |   |   TOK_INTCON "6" (7.13.19) int const
|   |   |   ';' ";" (7.13.21)
|   |   |   TOK_CALL "(" (7.14.11)
|   |   |   |   TOK_IDENT "putint" (7.14.4) variable
|   |   |   |   TOK_IDENT "a" (7.14.12) int variable (7.10.10)
|   |   |   TOK_CALL "(" (7.15.11)
|   |   |   |   TOK_IDENT "putchr" (7.15.4) variable
|   |   |   |   TOK_CHARCON "' '" (7.15.12) int const
|   |   |   TOK_CALL "(" (7.16.11)
|   |   |   |   TOK_IDENT "putint" (7.16.4) variable
|   |   |   |   TOK_ARROW "->" (7.16.13) lval vaddr
|   |   |   |   |   TOK_IDENT "b" (7.16.12) struct variable (7.11.22)
|   |   |   |   |   TOK_IDENT "a" (7.16.15) int variable (7.10.10)
|   |   |   TOK_CALL "(" (7.17.11)
|   |   |   |   TOK_IDENT "putchr" (7.17.4) variable
|   |   |   |   TOK_CHARCON "'\n'" (7.17.12) int const
|   |   |   TOK_RETURN "return" (7.18.4)
|   |   |   |   TOK_INTCON "0" (7.18.11) int const
