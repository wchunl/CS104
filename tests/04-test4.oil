fail:     .function
          .param ptr expr
          .param ptr file
          .param int line
          .end
putchr:   .function
          .param int chr
          .end
putint:   .function
          .param int num
          .end
putstr:   .function
          .param ptr str
          .end
getchr:   .function int
          .end
getstr:   .function
          .end
getln:    .function
          .end
          .struct foo
          .field int a
          .end
main:     .function int
          .local int a
          .local ptr b
          a = 6
          malloc (size of struct foo)
          b = $t0:i
          b->foo.a = 8
          $t1:i = * + 6
          a = $t1:i
          putint(a)
          putchr(' ')
          $t2:p = b->foo.a
          putint($t2:i)
          putchr('\n')
          return 0
          return 
          .end
