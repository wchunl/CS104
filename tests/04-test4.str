string_set[   0]:    4706206694200543916 0x1e7cdc8->","
                   16579916199511856337 0x1e7e488->"getln"
string_set[   1]:    2297668033614959926 0x1e80458->"0"
                    6637313742931709005 0x1e7f638->"*"
string_set[   3]:   11868120863286193964 0x1e7ec48->"main"
                    9611428142580219387 0x1e78758->"fail"
string_set[   4]:   10838281452030117757 0x1e7f318->"b"
                      21606623943142546 0x1e7d268->"{"
                    4017852018407011111 0x1e7ced8->"file"
string_set[   6]:   13686721134187162920 0x1e7ebd8->"}"
string_set[   8]:    1370730858159036685 0x1e7f768->"+"
                    9386170080689500517 0x1e7cc78->"expr"
string_set[   9]:   14526079891305879776 0x1e7f2a8->">"
string_set[  12]:   15816506932648033945 0x1e7cae8->"("
string_set[  16]:   13945322250941355849 0x1e7e6f8->"struct"
string_set[  17]:   17899137375351314596 0x1e7f618->"8"
                   17176659130743002837 0x1e7f038->"ptr"
                   18413295556186398326 0x1e7d2c8->"putint"
string_set[  18]:    7211130878751977441 0x1e7d418->"putchr"
                   12679825168387489523 0x1e7d9a8->"num"
string_set[  19]:    3835993668518112351 0x1e7ef88->"6"
string_set[  22]:   14805121035671447500 0x1e7f058->"->"
                    9631199822919835226 0x1e7e878->"foo"
string_set[  23]:   11166701219533395393 0x1e7e1f8->"getstr"
                   12868445110721718657 0x1e7cef8->")"
string_set[  25]:    4993892634952068459 0x1e7eb08->"a"
string_set[  26]:   18172887160629263681 0x1e7fac8->"' '"
string_set[  28]:   14635384904416252381 0x1e802f8->"return"
string_set[  29]:   12291392672316406109 0x1e7d0b8->"line"
string_set[  31]:    6192799629375248351 0x1e7ff48->"'\n'"
                    6142509188972423790 0x1e78738->""
                   13348923631108115385 0x1e7d738->"putstr"
string_set[  32]:   11597697714117577063 0x1e7eed8->"="
string_set[  33]:    8290819294854170237 0x1e7f368->"alloc"
                    1583576014053361350 0x1e7db58->"str"
string_set[  38]:   12597418120160561823 0x1e7ccf8->"int"
string_set[  39]:    7296690267450279874 0x1e7d588->"chr"
string_set[  40]:   11911796782241914524 0x1e78318->"void"
string_set[  42]:    5344662657899890615 0x1e7d0d8->";"
string_set[  44]:   13497578857207247412 0x1e7cb98->"string"
string_set[  45]:    1396094175042499165 0x1e7f138->"<"
string_set[  46]:    9201800968376702500 0x1e7df88->"getchr"
load_factor = 0.894
bucket_count = 47
max_bucket_size = 3
