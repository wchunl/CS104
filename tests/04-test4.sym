fail (6.22.1) {0} void function
   expr (6.22.12) {1} string variable param lval 0
   file (6.22.25) {1} string variable param lval 1
   line (6.22.38) {1} int variable param lval 2

putchr (6.24.1) {1} void function
   chr (6.24.14) {1} int variable param lval 0

putint (6.25.1) {2} void function
   num (6.25.14) {1} int variable param lval 0

putstr (6.26.1) {3} void function
   str (6.26.14) {1} string variable param lval 0

getchr (6.28.1) {4} int function

getstr (6.29.1) {5} string function

getln (6.30.1) {6} string function

foo (7.5.1) {0} struct foo
   a (7.6.4)  int field 0

main (7.9.1) {7} int function
   a (7.10.10) {1} int variable local lval 0
   b (7.11.22) {1} ptr <struct foo> variable local lval 1
