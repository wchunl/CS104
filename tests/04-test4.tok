#  1 "04-test4.oc"
#  1 "<built-in>"
#  1 "<command-line>"
#  1 "04-test4.oc"
#  1 "oclib.h"
# 22 "oclib.h"
   6 22.00001 261 TOK_VOID      void
   6 22.00006 282 TOK_IDENT     fail
   6 22.00011  40 '('           (
   6 22.00012 263 TOK_STRING    string
   6 22.00019 282 TOK_IDENT     expr
   6 22.00023  44 ','           ,
   6 22.00025 263 TOK_STRING    string
   6 22.00032 282 TOK_IDENT     file
   6 22.00036  44 ','           ,
   6 22.00038 262 TOK_INT       int
   6 22.00042 282 TOK_IDENT     line
   6 22.00046  41 ')'           )
   6 22.00047  59 ';'           ;
   6 24.00001 261 TOK_VOID      void
   6 24.00006 282 TOK_IDENT     putchr
   6 24.00013  40 '('           (
   6 24.00014 262 TOK_INT       int
   6 24.00018 282 TOK_IDENT     chr
   6 24.00021  41 ')'           )
   6 24.00022  59 ';'           ;
   6 25.00001 261 TOK_VOID      void
   6 25.00006 282 TOK_IDENT     putint
   6 25.00013  40 '('           (
   6 25.00014 262 TOK_INT       int
   6 25.00018 282 TOK_IDENT     num
   6 25.00021  41 ')'           )
   6 25.00022  59 ';'           ;
   6 26.00001 261 TOK_VOID      void
   6 26.00006 282 TOK_IDENT     putstr
   6 26.00013  40 '('           (
   6 26.00014 263 TOK_STRING    string
   6 26.00021 282 TOK_IDENT     str
   6 26.00024  41 ')'           )
   6 26.00025  59 ';'           ;
   6 28.00001 262 TOK_INT       int
   6 28.00005 282 TOK_IDENT     getchr
   6 28.00011  40 '('           (
   6 28.00012  41 ')'           )
   6 28.00013  59 ';'           ;
   6 29.00001 263 TOK_STRING    string
   6 29.00008 282 TOK_IDENT     getstr
   6 29.00014  40 '('           (
   6 29.00015  41 ')'           )
   6 29.00016  59 ';'           ;
   6 30.00001 263 TOK_STRING    string
   6 30.00008 282 TOK_IDENT     getln
   6 30.00013  40 '('           (
   6 30.00014  41 ')'           )
   6 30.00015  59 ';'           ;
#  4 "04-test4.oc"
   7 5.00001 269 TOK_STRUCT    struct
   7 5.00008 282 TOK_IDENT     foo
   7 5.00012 123 '{'           {
   7 6.00004 262 TOK_INT       int
   7 6.00008 282 TOK_IDENT     a
   7 6.00009  59 ';'           ;
   7 7.00001 125 '}'           }
   7 7.00002  59 ';'           ;
   7 9.00001 262 TOK_INT       int
   7 9.00005 282 TOK_IDENT     main
   7 9.00009  40 '('           (
   7 9.00010  41 ')'           )
   7 9.00012 123 '{'           {
   7 10.00004 262 TOK_INT       int
   7 10.00008 282 TOK_IDENT     a
   7 10.00010  61 '='           =
   7 10.00012 283 TOK_INTCON    6
   7 10.00013  59 ';'           ;
   7 11.00004 274 TOK_PTR       ptr
   7 11.00007  60 '<'           <
   7 11.00008 269 TOK_STRUCT    struct
   7 11.00015 282 TOK_IDENT     foo
   7 11.00018  62 '>'           >
   7 11.00020 282 TOK_IDENT     b
   7 11.00022  61 '='           =
   7 11.00024 273 TOK_ALLOC     alloc
   7 11.00029  60 '<'           <
   7 11.00030 269 TOK_STRUCT    struct
   7 11.00037 282 TOK_IDENT     foo
   7 11.00040  62 '>'           >
   7 11.00041  40 '('           (
   7 11.00042  41 ')'           )
   7 11.00043  59 ';'           ;
   7 12.00004 282 TOK_IDENT     b
   7 12.00005 272 TOK_ARROW     ->
   7 12.00007 282 TOK_IDENT     a
   7 12.00009  61 '='           =
   7 12.00011 283 TOK_INTCON    8
   7 12.00012  59 ';'           ;
   7 13.00004 282 TOK_IDENT     a
   7 13.00006  61 '='           =
   7 13.00008 282 TOK_IDENT     a
   7 13.00010  42 '*'           *
   7 13.00012 282 TOK_IDENT     b
   7 13.00013 272 TOK_ARROW     ->
   7 13.00015 282 TOK_IDENT     a
   7 13.00017  43 '+'           +
   7 13.00019 283 TOK_INTCON    6
   7 13.00020  59 ';'           ;
   7 13.00021  59 ';'           ;
   7 14.00004 282 TOK_IDENT     putint
   7 14.00011  40 '('           (
   7 14.00012 282 TOK_IDENT     a
   7 14.00013  41 ')'           )
   7 14.00014  59 ';'           ;
   7 15.00004 282 TOK_IDENT     putchr
   7 15.00011  40 '('           (
   7 15.00012 284 TOK_CHARCON   ' '
   7 15.00015  41 ')'           )
   7 15.00016  59 ';'           ;
   7 16.00004 282 TOK_IDENT     putint
   7 16.00011  40 '('           (
   7 16.00012 282 TOK_IDENT     b
   7 16.00013 272 TOK_ARROW     ->
   7 16.00015 282 TOK_IDENT     a
   7 16.00016  41 ')'           )
   7 16.00017  59 ';'           ;
   7 17.00004 282 TOK_IDENT     putchr
   7 17.00011  40 '('           (
   7 17.00012 284 TOK_CHARCON   '\n'
   7 17.00016  41 ')'           )
   7 17.00017  59 ';'           ;
   7 18.00004 268 TOK_RETURN    return
   7 18.00011 283 TOK_INTCON    0
   7 18.00012  59 ';'           ;
   7 19.00001 125 '}'           }
