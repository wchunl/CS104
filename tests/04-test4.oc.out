; 0.  1: # 1 "04-test4.oc"; 1.  0: 
; 1.  1: # 1 "<built-in>"
; 2.  1: # 1 "<command-line>"
; 3.  1: # 1 "04-test4.oc"
; 4.  1: 
; 4.  2: 
; 4.  3: # 1 "oclib.h" 1
; 5.  1: # 22 "oclib.h"
; 6. 22: void fail (string expr, string file, int line);
; 6. 23: 
; 6. 24: void putchr (int chr);
; 6. 25: void putint (int num);
; 6. 26: void putstr (string str);
; 6. 27: 
; 6. 28: int getchr();
; 6. 29: string getstr();
; 6. 30: string getln();
; 6. 31: # 4 "04-test4.oc" 2
; 7.  4: 
; 7.  5: struct foo {
; 7.  6:    int a;
; 7.  7: };
; 7.  8: 
; 7.  9: int main() {
; 7. 10:    int a = 6;
; 7. 11:    ptr<struct foo> b = alloc<struct foo>();
; 7. 12:    b->a = 8;
; 7. 13:    a = a * b->a + 6;;
; 7. 14:    putint (a);
; 7. 15:    putchr (' ');
; 7. 16:    putint (b->a);
; 7. 17:    putchr ('\n');
; 7. 18:    return 0;
; 7. 19: }
